File: src/algorithms/python/Plus_One.py
# -*- coding: utf-8 -*-


class Solution:
    # @param {integer[]} digits
    # @return {integer[]}
    def plusOne(self, digits):
        left, ret = 0, []
        digits[-1] += 1
        for index in range(len(digits) - 1, -1, -1):
            digits[index] += left
            left = 1 if digits[index] >= 10 else 0
            digits[index] %= 10
            ret = [digits[index]] + ret
        if left:
            ret = [left] + ret
        return ret


if __name__ == '__main__':
    solution = Solution()
    print solution.plusOne([1, 2, 3, 9])
    print solution.plusOne([9, 9])
    print solution.plusOne([1, 2, 3])

File: src/algorithms/cpp/Plus_One.cpp
#include <iostream>
#include <cstdio>
#include <cstring>
#include <cmath>
#include <cstdlib>
#include <algorithm>
#include <vector>
#include <map>
#include <set>
using namespace std;
#define out(v) cerr << #v << ": " << (v) << endl
#define SZ(v) ((int)(v).size())
const int maxint = -1u>>1;
template <class T> bool get_max(T& a, const T &b) {return b > a? a = b, 1: 0;}
template <class T> bool get_min(T& a, const T &b) {return b < a? a = b, 1: 0;}

class Solution {
public:
    vector <int> plusOne(vector <int>& digits) {
        int left = 1;
        for (int i = SZ(digits) - 1; i >= 0; -- i) {
            digits[i] += left;
            left = digits[i] / 10;
            digits[i] %= 10;
        }
        if (left)
            digits.insert(digits.begin(), left);
        return digits;
    }
};

int main() {
    vector <int> v;
    v.push_back(9), v.push_back(9);
    Solution solution = Solution();
    vector <int> result = solution.plusOne(v);
    for (int i = 0; i < SZ(result); ++ i) {
        cout << result[i] << endl;
    }
    return 0;
}
